File: 0347/347.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
 var topKFrequent = function(nums, k) {
    
    //initialize
    var ret = []


    var arr = [];

    //sort the array O(n log(n))
    nums.sort((a,b)=>{
        return a-b;
    })

    var nowNum = nums[0];
    
    var nowCount = 1;

    //loop for make the array;
    for(var pos = 1;pos<=nums.length;pos++)
    {
        if(nowNum!=nums[pos]||pos ==nums.length )
        {
            var param = [nowNum,nowCount]
            arr.push(param)

            //reset everything
            nowNum = nums[pos]
            nowCount =1;

        }else
        {
            //continue to count
            nowCount++;
        }
    }

    arr.sort((a,b)=>{
        return b[1]-a[1]
    })

    for(var i = 0;i<k;i++)
    {
        ret.push(arr[i][0])
    }

    return ret;
};

var nums=[5,2,5,3,5,3,1,1,3]
var k = 2
console.log(topKFrequent(nums,k))

File: 0347/347.cpp
class Solution {
public:
    struct intArr
    {
        int num;
        int count;
    };
    static bool sort_func(intArr a, intArr b)
    {
        //cout<<a.count<<b.count<<"\n";
        return !(a.count<b.count);
    }
    
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> ret;
        vector<struct intArr> save;
        int curNum = 0,curCount =0;
        
        //sort the nums
        sort(nums.begin(), nums.end());
        
        //init first 
        curNum = nums[0];
        curCount = 1;
        
        //loop all count 
        for(int i = 1;i<nums.size();i++)
        {
            if(nums[i] != curNum )
            {
                struct intArr addi;
                addi.num = curNum;
                addi.count = curCount;
                save.push_back(addi);
                
                curNum = nums[i];
                curCount = 1;
            }else
            {
                curCount++;
            }
        }
        //last one
        struct intArr addi;
        addi.num = curNum;
        addi.count = curCount;
        save.push_back(addi);
        
        
        //sort the vector for checking 
        sort(save.begin(),save.end(),sort_func);
        
        //make ans vector
        for(int i = 0;i<k;i++)
        {
            //cout<<save[i].num<<"\n";
            ret.push_back(save[i].num);
        }
        
        return ret;
    }
};